src/builtins.c

```diff
@@ -270,8 +270,15 @@ JL_CALLABLE(jl_f_sizeof)
             jl_error("type does not have a fixed size");
         return jl_box_long(jl_datatype_size(x));
     }
-    if (jl_is_array(x))
-        return jl_box_long(jl_array_len(x) * ((jl_array_t*)x)->elsize);
+    if (jl_is_array(x)) {
+        jl_value_t *ety = jl_tparam0(jl_typeof(x));
+        size_t elsize = 0, al = 0;
+        int isboxed = !jl_islayout_inline(ety, &elsize, &al);
+        if (isboxed)
+            elsize = sizeof(void*);
+        size_t len = jl_array_len(x);
+        return jl_box_long((!isboxed && jl_is_uniontype(ety)) ? len * elsize + len : len * elsize);
+    }
```

Comment:
You're right, we can just use `->elsize`. I updated the PR to *not* include the `+ len` as part of the size. I think it will keep this more on the "optimization" side of things as opposed to necessarily committing to the layout. I also updated to give a size for isbits Union types.